packages/model-generator/src/create_graphql_models_generator.ts

```diff
@@ -34,78 +48,80 @@ export const createGraphqlModelsGenerator = (
 
 export type GraphqlModelsFromBackendIdentifierParams = {
   backendIdentifier: DeployedBackendIdentifier;
-  credentialProvider: AwsCredentialIdentityProvider;
+  options: GraphqlModelsFetchOptions;
 };
 
 /**
  * Factory function to compose a model generator from a backend identifier.
  */
 const createGraphqlModelsGeneratorFromBackendIdentifier = ({
   backendIdentifier,
-  credentialProvider,
+  options,
 }: GraphqlModelsFromBackendIdentifierParams): GraphqlModelsGenerator => {
   if (!backendIdentifier) {
     throw new Error('`backendIdentifier` must be defined');
   }
-  if (!credentialProvider) {
-    throw new Error('`credentialProvider` must be defined');
+  if (!options) {
+    throw new Error('`options` must be defined');
```

Comment:
Name this variable `option` something more meaningful throughout this PR, instead of this generic name which leads to a dumping group for anything and everything.

Also getting this error `options` is not defined gives customer or maintainers no context on what went wrong.

Reply:
Do you have a recommendation? Maybe `clientOptions` or `credentialsOptions`? I tried to use as much prior art as possible, the name `options` came from the similar implementation from https://github.com/aws-amplify/amplify-backend/blob/main/packages/deployed-backend-client/src/backend_output_client_factory.ts#L59 and the error message `${paramName} must be defined` is from the line above this one:

```
  if (!backendIdentifier) {
    throw new Error('`backendIdentifier` must be defined');
  }
```

Reply:
I think we would need three clients - cfn, amplify, and s3. The reason I think it's better to combine them into one object is that the validation is done for us as build time instead of runtime and it mirrors the api for `deployed-backend-client. Having a bunch of optional params can be confusing about when they're required and when they're not, but having well defined and logical types it's a lot more clear.

Reply:
>it mirrors the api for deployed-backend-client so it provides consistency across packages

Instead we should try to fix the issues across packages to provide consistency.

1. Looks like we should have an `AWSClientsFactory`? Which can create clients from
   1. Preconfigured clients
   2. Passed in credentials
   3. Default credentials
2. If both clients and credentials do belong in a type, then we need to name it accordingly (based on what they represent or what do they have in common). `Option` basically means an open ended collection of parameters

Reply:
Would it always create new cfn, amplify, and s3 clients regardless of what clients are needed by the consumer? If some module requires a fourth client, would it be added to all consumers of any aws client?

Reply:
You can create a `ClientProvider` and configure it from browser with pre-configured clients or however you want and the node would configure it with credentials (either custom or default) and pass this `clientProvider` to the commands. Commands can then do `clientProvider.getS3Client` and so on.

Reply:
I like this idea - let me give it a try and see how it feels.